client/state/test/data-fetching-persistence.tsx

```diff
@@ -0,0 +1,281 @@
+	error: noop,
+	log: noop,
+	warn: noop,
+} );
+
+class Storage {
+	cache: Map< string, string >;
+
+	constructor() {
+		this.cache = new Map();
+	}
+
+	getItem( key: string ) {
+		return this.cache.get( key );
+	}
+
+	setItem( key: string, value: string ) {
+		this.cache.set( key, value );
+	}
+
+	removeItem( key: string ) {
+		this.cache.delete( key );
+	}
+
+	clear() {
+		this.cache.clear();
+	}
+
+	get length() {
+		return this.cache.size;
+	}
+
+	key() {
+		return 'noop';
+	}
+}
+
+const storage = new Storage();
+
+const getOfflinePersistence = () => JSON.parse( storage.getItem( PERSISTENCE_KEY ) );
+
+const offlinePersistor = createWebStoragePersistor( {
+	storage,
+	key: PERSISTENCE_KEY,
+	throttleTime: 0,
+} );
+
+interface DataFetchingComponentProps< T > {
+	queryFn(): Promise< T >;
+	persistencePredicate?: boolean | ( ( data: T ) => boolean );
+}
+
+const DataFetchingComponent = < T, >( {
+	queryFn,
+	persistencePredicate,
+}: DataFetchingComponentProps< T > ) => {
+	const { status } = useQuery( queryKey, queryFn, {
+		retry: false,
+		meta: {
+			persist: persistencePredicate != null ? persistencePredicate : undefined,
+		},
+	} );
+
+	return <>{ status }</>;
+};
+
+const TestComponent = < T, >( dataFetchingProps: DataFetchingComponentProps< T > ) => {
+	return (
+		<QueryClientProvider client={ queryClient }>
+			<DataFetchingComponent { ...dataFetchingProps } />
+		</QueryClientProvider>
+	);
+};
+
+describe( 'shouldDehydrateQuery', () => {
+	beforeAll( async () => {
+		await persistQueryClient( {
+			queryClient,
+			persistor: offlinePersistor,
+			dehydrateOptions: {
+				shouldDehydrateQuery,
+			},
+		} );
+	} );
+
+	beforeEach( () => {
+		storage.clear();
+		queryClient.clear();
+	} );
+
+	it( 'does not persist failed queries', async () => {
+		const data = 'Hello, World!';
+
+		const queryFn = () => Promise.reject( data );
+
+		const { getByText } = render(
+			<TestComponent queryFn={ queryFn } persistencePredicate={ true } />
+		);
+
+		await waitFor( () => getByText( 'error' ) );
+
+		const cache = await waitFor( () => getOfflinePersistence() );
+
+		expect( cache ).toEqual(
+			expect.objectContaining( {
+				buster: '',
+				timestamp: expect.any( Number ),
+				clientState: {
+					mutations: [],
+					queries: [],
+				},
+			} )
+		);
+	} );
+
+	describe( 'when passing `true` to `shouldPersistQuery`', () => {
+		it( 'persists the query', async () => {
+			const data = 'Hello, World!';
+
+			const queryFn = () => Promise.resolve( data );
+
+			const { getByText } = render(
+				<TestComponent queryFn={ queryFn } persistencePredicate={ true } />
+			);
+
+			await waitFor( () => getByText( 'success' ) );
+
+			const cache = await waitFor( () => getOfflinePersistence() );
+
+			expect( cache ).toEqual(
+				expect.objectContaining( {
+					buster: '',
+					timestamp: expect.any( Number ),
+					clientState: {
+						mutations: [],
+						queries: [
+							{
+								state: {
+									data: 'Hello, World!',
```

Comment:
We could shorten this by reusing the `data` constant:

```suggestion
									data,
```